lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -114,7 +114,8 @@ static SILLinkage getAutoDiffFunctionLinkage(SILLinkage originalLinkage,
   // associated function. Make the associated function public unless
   // differentiation is not explicitly requested.
   if (originalLinkage == SILLinkage::Public ||
-      originalLinkage == SILLinkage::PublicNonABI)
+      originalLinkage == SILLinkage::PublicNonABI ||
+      originalLinkage == SILLinkage::Shared)
```

Comment:
In the example of `differentiableFunction(from:)`, the original function is:

```
// original #1 <A, B>(_:) in differentiableFunction<A, B>(from:)
sil shared [serializable] [differentiable source 0 wrt 0 vjp @$s3dif22differentiableFunction4fromq_xcq_5value_15CotangentVectorQzAEQy_c8pullbacktxc_ts14DifferentiableRzsAIR_r0_lF10derivativeL_yq_AD_AfGcAHtxsAIRzsAIR_r0_lF] [ossa] @$s3dif22differentiableFunction4fromq_xcq_5value_15CotangentVectorQzAEQy_c8pullbacktxc_ts14DifferentiableRzsAIR_r0_lF8originalL_yq_xsAIRzsAIR_r0_lF : $@convention(thin) <T, R where T : Differentiable, R : Differentiable> (@in_guaranteed T, @guaranteed @callee_guaranteed (@in_guaranteed T) -> (@out R, @owned @callee_guaranteed (@in_guaranteed R.CotangentVector) -> @out T.CotangentVector)) -> @out R
```

A more precise condition here might be `originalLinkage == SILLinkage::Shared && original->isSerialized()`.
`original->isSerialized()` returns true when `original` is `serialized` or `serializable`.

Reply:
Seems like the right fix. I think that `PublicNonABI` can even be removed here, but let's figure that out later.